server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsInProgress.java

```diff
@@ -76,13 +92,52 @@ public SnapshotDeletionsInProgress withAddedEntry(Entry entry) {
     }
 
     /**
-     * Returns a new instance of {@link SnapshotDeletionsInProgress} which removes
-     * the given entry from the invoking instance.
+     * Returns a new instance of {@link SnapshotDeletionsInProgress} that has the entry with the given {@code deleteUUID} removed from its
+     * entries.
      */
-    public SnapshotDeletionsInProgress withRemovedEntry(Entry entry) {
-        List<Entry> entries = new ArrayList<>(getEntries());
-        entries.remove(entry);
-        return new SnapshotDeletionsInProgress(entries);
+    public SnapshotDeletionsInProgress withRemovedEntry(String deleteUUID) {
+        List<Entry> updatedEntries = new ArrayList<>(entries.size() - 1);
+        boolean removed = false;
+        for (Entry entry : entries) {
+            if (entry.uuid().equals(deleteUUID)) {
+                removed = true;
+            } else {
+                updatedEntries.add(entry);
+            }
+        }
+        return removed ? new SnapshotDeletionsInProgress(updatedEntries) : this;
+    }
+
+    public SnapshotDeletionsInProgress withRemovedSnapshotIds(String repository, Collection<SnapshotId> snapshotIds) {
```

Comment:
If two or more deletes share overlapping snapshots but couldn't be batched into one we need to "clean up" deleted snapshots from existing deletes so that the remaining snapshots don't fail trying to delete already deleted snapshots.